src/dawn/IIR/StatementAccessesPairIterator.h

```diff
@@ -0,0 +1,131 @@
+﻿//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#ifndef DAWN_IIR_SAPITERATOR_H
+#define DAWN_IIR_SAPITERATOR_H
+
+#include "dawn/IIR/IIRNode.h"
+#include "dawn/IIR/IIRNodeIterator.h"
+#include "dawn/IIR/StatementAccessesPair.h"
+#include "dawn/SIR/ASTNodeIterator.h"
+#include "dawn/Support/IteratorRange.h"
+#include <unordered_map>
+
+namespace dawn {
+
+/// @brief Iterator through StatementAccessesPairs of statements of an AST
+///
+/// Built on top of an ASTNodeIterator that iterates through statements, requires a map from ASTStmt
+/// to StatementAccessesPair. Can either visit the first level of the AST or the whole AST.
+template <ASTNodeIteratorVisitKind onlyFirstLevel>
+class StatementAccessesPairIterator {
+public:
+  using WrappedIterator = ASTNodeIterator<onlyFirstLevel>;
+  using ASTStmtToSAPMapType = std::unordered_map<const Stmt*, iir::StatementAccessesPair>;
+
+  StatementAccessesPairIterator(const std::shared_ptr<Stmt>& root,
+                                const ASTStmtToSAPMapType& ASTStmtToSAPMap)
+      : current_(WrappedIterator::CreateInstance(root)),
+        ASTStmtToSAPMap_(const_cast<ASTStmtToSAPMapType&>(ASTStmtToSAPMap)) {}
+  StatementAccessesPairIterator(const WrappedIterator& it,
+                                const ASTStmtToSAPMapType& ASTStmtToSAPMap)
+      : current_(it.clone()), ASTStmtToSAPMap_(const_cast<ASTStmtToSAPMapType&>(ASTStmtToSAPMap)) {}
+  StatementAccessesPairIterator(StatementAccessesPairIterator&& it) = default;
+  StatementAccessesPairIterator(const StatementAccessesPairIterator&) = delete;
+  StatementAccessesPairIterator& operator=(StatementAccessesPairIterator&& o) {
+    *this = std::move(o);
+  }
+  StatementAccessesPairIterator& operator=(const StatementAccessesPairIterator&) = delete;
+
+  StatementAccessesPairIterator clone() const {
+    return StatementAccessesPairIterator(current_, ASTStmtToSAPMap_);
```

Comment:
this gives you a shallow copy of the map

Reply:
Doesn't it initialize the reference `ASTStmtToSAPMap_` to point to the original map? we don't want to copy the map, just to reference the original one